ls1intum/Artemis
# `Plagiarism checks`: Clean up results and cases in case of errors

### Checklist
#### General
#### Server

### Motivation and Context
The cpc leaves phantom plagiarism cases when any of its subsequent runs results in an error. Such phantom cases can lead to false positive plagiarism case notification or not removing plagiarism case after cpc detected no plagiarism. An example scenario is when there suddenly is no submissions matching the plagiarism check configuration.

### Description
Now in case an exception Is thrown the cpc will remove all phantom plagiarism comparisons and cases, leaving the system in a correct state.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Text Exercise
- ⚠️ deployed brand which triggers cpc every 1 minute: `plagiarism-detection-fix-cpc-when-new-short-submission-1min` ⚠️ 

1. Create an exercise (text, modeling or programming) with cpc enabled
    - set a future due date for the exercise
    - check the _Enable Continuos plagiarism control_ checkbox
3. Submit two identical submissions from two students. Each submission needs to be at least 200 words long.
4. Wait 1min until the plagiarism case is created.
5. Verify that plagiarism cases was created for both students.
6. Change one of submissions to a 2-3 words long text.
4. Wait 1min.
5. Verify that both plagiarism case are removed.



### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress
#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ContinuousPlagiarismControlService | 96% | ✅                           |
